tiledb/sm/filter/bitsort_filter.cc

```diff
@@ -0,0 +1,465 @@
+
+Status BitSortFilter::run_forward(
+    const Tile& tile,
+    BitSortFilterMetadataType &pair,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+
+  // Since run_forward interprets the filter's data as integers, we case on
+  // the size of the type and pass in the corresponding integer type into
+  // a templated function.
+  auto tile_type = tile.type();
+  switch (datatype_size(tile_type)) {
+    case sizeof(uint8_t): {
+      return run_forward<uint8_t>(
+          pair, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint16_t): {
+      return run_forward<uint16_t>(
+         pair, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint32_t): {
+      return run_forward<uint32_t>(
+          pair, input_metadata, input, output_metadata, output);
+    }
+    case sizeof(uint64_t): {
+      return run_forward<uint64_t>(
+          pair, input_metadata, input, output_metadata, output);
+    }
+    default: {
+      return Status_FilterError(
+          "BitSortFilter::run_forward: datatype does not have an appropriate "
+          "size");
+    }
+  } 
+
+  return Status_FilterError("BitSortFilter::run_forward: invalid datatype.");
+}
+
+template <typename T>
+Status BitSortFilter::run_forward(
+    BitSortFilterMetadataType &pair,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  // Output size does not change with this filter.
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Write the metadata.
+  auto parts = input->buffers();
+  auto num_parts = (uint32_t)parts.size();
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // Keep track of the starting offsets to write in positions for each part.
+  std::vector<uint32_t> offsets;
+  uint32_t total_size = 0;
+  for (const auto& part : parts) {
+    auto part_size = (uint32_t)part.size();
+    offsets.push_back(total_size);
+    total_size += part_size;
+  }
+
+  // Sort all parts.
+  std::vector<std::pair<T, uint64_t>> sorted_elements(total_size / sizeof(T));
+  for (uint64_t i = 0; i < num_parts; ++i) {
+    const auto &part = parts[i];
+    auto part_size = static_cast<uint32_t>(part.size());
+    RETURN_NOT_OK(output_metadata->write(&part_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(sort_part<T>(&part, output_buf, offsets[i], sorted_elements));
+  }
+
+  // Rewrite each of the dimension tile data with the new sort order.
+  std::vector<Tile*> &dim_tiles = pair.first.get();
+
+  // Since rewrite_dim_tile_forward is only moving around dimension data and 
```

Comment:
Depending on if we can remove unnecessary copies in this code, we might want to rename this. I think there is a possibility to directly move the data from the filtered buffer to the unfiltered buffer for the dimensions, so in this case, this should be named run_forward_dim_tile.